### PR TITLE
Move `Memory` into `MemorySegmentManager`

### PR DESCRIPTION
**Pull request by fmoletta**
This PR aims to reduce the complexity of the VM's memory. and also enforce proper usage (as the memory and its segment manager are now a "unified" entity)
- Remove `memory` field from `VirtualMachine`
- Add `memory` field to `MemorySegmentManager`
- Remove `Memory` argument from methods where `MemorySegmentManager` is also an argument
- Add test macro `segments` (an extension of the `memory` macro

### CODE REVIEW

File: src/vm/runners/builtin_runner/keccak.rs [hunk 1]
```diff
@@ -263,11 +259,13 @@ mod tests {
     use crate::types::program::Program;
     use crate::utils::test_utils::*;
     use crate::vm::runners::cairo_runner::CairoRunner;
+    use crate::vm::vm_memory::memory::Memory;
     use crate::vm::{
         errors::{memory_errors::MemoryError, runner_errors::RunnerError},
         runners::builtin_runner::BuiltinRunner,
         vm_core::VirtualMachine,
     };
+    use std::collections::HashMap;
```
Comment (Jrigada) [inline src/vm/runners/builtin_runner/keccak.rs:268]: Why is this line new? Where are we using Hashmap?

**Review by Jrigada — APPROVED** [review]

File: src/vm/vm_memory/memory_segments.rs [hunk 2]
```diff
@@ -13,23 +13,23 @@ use std::{
     collections::{HashMap, HashSet},
 };
 
-#[derive(Debug, PartialEq, Eq)]
 pub struct MemorySegmentManager {
     pub num_segments: usize,
     pub num_temp_segments: usize,
     pub segment_sizes: HashMap<usize, usize>,
     pub segment_used_sizes: Option<Vec<usize>>,
+    pub memory: Memory,
```
Comment (Oppen) [inline src/vm/vm_memory/memory_segments.rs:21]: Please `pub(crate)`.

File: src/utils.rs [hunk 3]
```diff
@@ -102,6 +102,26 @@ pub mod test_utils {
...
+    macro_rules! segments {
+        ($( (($si:expr, $off:expr), $val:tt) ),* ) => {
+            {
+                let memory = memory!($( (($si, $off), $val) ),*);
+                let mem_len = memory.data.len();
+                MemorySegmentManager {
+                    memory,
+                    num_segments: mem_len,
+                    num_temp_segments: 0,
+                    segment_sizes: HashMap::new(),
```
[the remark this answers is not in the document]
Comment (fmoletta) [reply src/utils.rs:114]: On it!

[the remark this answers is not in the document]
Comment (fmoletta) [reply src/utils.rs:111]: `segment_sizes` is not calculated by taking the lenght of a segment, it is written when calling `MemorySegmentManager.finalize()`

[the remark this answers is not in the document]
Comment (fmoletta) [reply src/utils.rs:111]: I also tried to apply this to `segment_used_sizes`, but again, this isnt just the lenght as it doesnt take into account memory gaps. This also led me to find a bug, as we are currently using the segment's lenght for segment_used_size which is incorrect

[the remark this answers is not in the document]
Comment (fmoletta) [reply src/utils.rs:111]: I realized I misread this review

[the remark this answers is not in the document]
Comment (fmoletta) [reply src/utils.rs:111]: I realized I misread the review, but it was helpful nonetheless

[the remark this answers is not in the document]
Comment (Oppen) [reply src/utils.rs:111]: > segment_sizes is not calculated by taking the lenght of a segment, it is written when calling MemorySegmentManager.finalize()
>
> I also tried to apply this to segment_used_sizes, but again, this isnt just the lenght as it doesnt take into account memory gaps (aka, segment_used_sizes doesnt count them) . This also led me to find a bug, as we are currently using the segment's lenght for segment_used_size which is incorrect

Oh, I thought only `segment_used_sizes` had that extra computation.

**Review by Oppen — APPROVED** [review]

File: src/utils.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (fmoletta) [reply src/utils.rs:111]: Update on this: it is indeed the segment lengh, the method's documentation( in cairo-rs) is incorrect

**Review by Oppen — APPROVED** [review]

**Review by Oppen — APPROVED** [review]

**Pull request merged**